### PR TITLE
[SYCL] Support building sycl runtime with libcxx in configure script

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: gejin <ge.jin@intel.com>
Currently, we claim to support building sycl runtime with llvm libc++ as an experimental feature but building with libcxx will fail in latest sources. This patch includes:
1. Adding support in configure script to build sycl runtime with llvm libc++. We need to manually construct the cmake commands previously.
2. Fix some building failures:
disabling xpti module when building sycl runtime with libcxx temporarily
use libcxx to compile object files required by sycl runtime library during link time.

### CODE REVIEW

File: sycl/CMakeLists.txt [hunk 1]
```diff
@@ -220,20 +220,28 @@ if (MSVC)
...
 
+function(add_libcxx_compile_options LIB_NAME)
+  target_compile_options(${LIB_NAME} PRIVATE -nostdinc++)
+  target_include_directories(${LIB_NAME} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
+endfunction(add_libcxx_include_option)
+
+function(add_libcxx_link_options LIB_NAME)
+  target_link_libraries(${LIB_NAME} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
+endfunction(add_libcxx_link_options)
+
```
Comment (alexbatashev) [inline sycl/CMakeLists.txt:231]: I think, this file becomes quite a mess. It's hard to read and navigate here. How about refactoring this to something like `cmake/modules/SYCLUtils.cmake`? Not a very strong suggestion, though.

@bader, @vladimirlaz  what do you think?

File: buildbot/configure.py [hunk 2]
```diff
@@ -101,6 +109,15 @@ def do_configure(args):
     # Add path to root CMakeLists.txt
     cmake_cmd.append(llvm_dir)
 
+    if args.use_libcxx:
+      if not (args.libcxx_include and args.libcxx_library):
+        sys.exit("Please specify include and library path of libc++ when building sycl "
+                 "runtime with it")
```
Comment (alexbatashev) [inline buildbot/configure.py:115]: Is it a strong requirement? Can we fallback to libc++, that resides in this source tree?

**Review by alexbatashev — CHANGES_REQUESTED** [review]
Overall, LGTM. The other two comments are minor nits, and can be handled separately.

What really bothers me, is that this patch breaks a few ABI-related lits. @jinge90 would you mind adding a new feature (like `libcxx` or something, see https://github.com/intel/llvm/blob/5cf8088c994778561c8584d5433d7d32618725b2/sycl/test/lit.cfg.py#L43 for example), and mark tests from `sycl/tests/abi` as unsupported, when that feature is available?

**Comment by jinge90** [comment]
> Overall, LGTM. The other two comments are minor nits, and can be handled separately.
> 
> What really bothers me, is that this patch breaks a few ABI-related lits. @jinge90 would you mind adding a new feature (like `libcxx` or something, see
> 
> https://github.com/intel/llvm/blob/5cf8088c994778561c8584d5433d7d32618725b2/sycl/test/lit.cfg.py#L43
> 
> for example), and mark tests from `sycl/tests/abi` as unsupported, when that feature is available?

Hi, @alexbatashev 
Thanks for your heuristic suggestion! I think we are faced with a more general question: "do we need to support sycl lit tests for sycl runtime built with libc++".
If we use compiler + sycl runtime(built with llvm libc++) to run all tests in llvm/sycl/test, all of them will report link time error.
The reason is we need to use llvm libc++ to build sycl applications when the sycl runtime is built with libc++ and using llvm libc++ requires some compiler options: https://libcxx.llvm.org/docs/UsingLibcxx.html
For example, to run a test using normal compiler/runtime, we may use:
clang++ -fsycl test.cpp -o test.out
./test.out

To run with compiler/runtime(built with llvm libc++), we need to use:
clang++ -fsycl -stdlib=libc++ -nostdinc++ test.cpp -I <path to llvm libc++ include headers> -L <library path of llvm libc++ libraries> -lc++ -lc++abi -o test.out
./test.out

If we want to run sycl tests using sycl runtime built with llvm libc++, it seems we need to update all sycl test "build command" to add support for llvm libc++ which is not a small work.
I haven't run sycl tests with runtime built with libc++ but expect most of them should pass. Some of the tests depend on GNU libstdc++ such as abi/layout_* and some devicelib tests, we may need to disable or support.
So, do we need to support sycl test for sycl runtime built with libc++?

**Comment by vladimirlaz** [comment]
> If we use compiler + sycl runtime(built with llvm libc++) to run all tests in llvm/sycl/test, all of them will report link time error.
> The reason is we need to use llvm libc++ to build sycl applications when the sycl runtime is built with libc++ and using llvm libc++ requires some compiler options: https://libcxx.llvm.org/docs/UsingLibcxx.html
> For example, to run a test using normal compiler/runtime, we may use:
> clang++ -fsycl test.cpp -o test.out
> ./test.out
> 
> To run with compiler/runtime(built with llvm libc++), we need to use:
> clang++ -fsycl -stdlib=libc++ -nostdinc++ test.cpp -I <path to llvm libc++ include headers> -L <library path of llvm libc++ libraries> -lc++ -lc++abi -o test.out
> ./test.out
> 
As soon as you need to change command line for all tests it is reasonable to modify @clangxx substitution by adding required command line arguments (when compiler is build with llvm libc++).

@mdtoguchi can/should be modify command line by clang driver itself to add required command line option implicitly when compiler is built with llvm libc++?

**Comment by jinge90** [comment]
> > If we use compiler + sycl runtime(built with llvm libc++) to run all tests in llvm/sycl/test, all of them will report link time error.
> > The reason is we need to use llvm libc++ to build sycl applications when the sycl runtime is built with libc++ and using llvm libc++ requires some compiler options: https://libcxx.llvm.org/docs/UsingLibcxx.html
> > For example, to run a test using normal compiler/runtime, we may use:
> > clang++ -fsycl test.cpp -o test.out
> > ./test.out
> > To run with compiler/runtime(built with llvm libc++), we need to use:
> > clang++ -fsycl -stdlib=libc++ -nostdinc++ test.cpp -I <path to llvm libc++ include headers> -L <library path of llvm libc++ libraries> -lc++ -lc++abi -o test.out
> > ./test.out
> 
> As soon as you need to change command line for all tests it is reasonable to modify @clangxx substitution by adding required command line arguments (when compiler is build with llvm libc++).
> 
> @mdtoguchi can/should be modify command line by clang driver itself to add required command line option implicitly when compiler is built with llvm libc++? I assume the library will be shipped in compiler directory structure in this case.

Hi, @vladimirlaz 
I am afraid we still need some option to indicate compiler that the application will run with sycl runtime built with libc++ since we won't know which sycl runtime will be used in compilation time even if we ship libc++/libc++abi with compiler in the future.

File: buildbot/configure.py [hunk 3]
```diff
@@ -21,17 +21,25 @@ def do_configure(args):
     ocl_header_dir = os.path.join(abs_obj_dir, "OpenCL-Headers")
     icd_loader_lib = os.path.join(abs_obj_dir, "OpenCL-ICD-Loader", "build")
     llvm_targets_to_build = 'X86'
-    llvm_enable_projects = 'clang;llvm-spirv;sycl;opencl-aot;xpti;libdevice'
+    llvm_external_projects = 'sycl;llvm-spirv;opencl-aot;libdevice'
+    llvm_enable_projects = 'clang;llvm-spirv;sycl;opencl-aot;libdevice'
```
Comment (bader) [inline buildbot/configure.py:25]: Please, re-use `llvm_external_projects` value here.

File: buildbot/configure.py [hunk 4]
```diff
@@ -21,17 +21,25 @@ def do_configure(args):
...
     libclc_targets_to_build = ''
     sycl_build_pi_cuda = 'OFF'
     sycl_werror = 'ON'
     llvm_enable_assertions = 'ON'
     llvm_enable_doxygen = 'OFF'
     llvm_enable_sphinx = 'OFF'
     llvm_build_shared_libs = 'OFF'
+    sycl_enable_xpti_tracing = 'ON'
```
Comment (bader) [inline buildbot/configure.py:33]:
```suggestion
    sycl_enable_xpti_tracing = 'OFF' if args.use_libcxx else 'ON'
```

File: buildbot/configure.py [hunk 5]
```diff
@@ -21,17 +21,25 @@ def do_configure(args):
...
 
     icd_loader_lib = os.path.join(icd_loader_lib, "libOpenCL.so" if platform.system() == 'Linux' else "OpenCL.lib")
 
+    if not args.use_libcxx:
+        llvm_enable_projects += ';xpti'
+        llvm_external_projects += ';xpti'
```
Comment (bader) [inline buildbot/configure.py:39]: I suggest moving this code to lines 24-25 - closer to variables initialization.

I also suggest moving sycl_enable_xpti_tracing initialization before llvm_enable_projects and llvm_external_projects  to re-use it for variables initialization.

Comment (jinge90) [reply buildbot/configure.py]: Done.

File: buildbot/configure.py (hunk 4 above)
Comment (jinge90) [reply buildbot/configure.py]: Done.

File: buildbot/configure.py (hunk 3 above)
Comment (jinge90) [reply buildbot/configure.py]: Done.

File: sycl/CMakeLists.txt (hunk 1 above)
Comment (jinge90) [reply sycl/CMakeLists.txt:231]: Hi, @alexbatashev 
I think the refactoring work can be done in a separate PR, is it OK?
Thanks very much.

File: buildbot/configure.py (hunk 2 above)
Comment (jinge90) [reply buildbot/configure.py:115]: If we need to fallback to libc++, we introduce the dependency of libcxx which means we need to build libc++ and libc++abi. Since building sycl runtime with libcxx is experimental feature, requiring users to provide a stable pre-built libcxx simplifies our work.
Thanks very much.

File: sycl/CMakeLists.txt (hunk 1 above)
Comment (alexbatashev) [reply sycl/CMakeLists.txt:231]: Yes, that's right

File: buildbot/configure.py (hunk 2 above)
Comment (alexbatashev) [reply buildbot/configure.py:115]: Ok, agreed

**Comment by jinge90** [comment]
> Overall, LGTM. The other two comments are minor nits, and can be handled separately.
> 
> What really bothers me, is that this patch breaks a few ABI-related lits. @jinge90 would you mind adding a new feature (like `libcxx` or something, see
> 
> https://github.com/intel/llvm/blob/5cf8088c994778561c8584d5433d7d32618725b2/sycl/test/lit.cfg.py#L43
> 
> for example), and mark tests from `sycl/tests/abi` as unsupported, when that feature is available?

Hi, @alexbatashev 
Just added "libc++" feature and marked all tests from sycl/test/abi as "UNSUPPORTED: libc++".
Thanks very much.

File: sycl/test/lit.cfg.py [hunk 6]
```diff
@@ -41,6 +41,7 @@
 # Configure LD_LIBRARY_PATH or corresponding os-specific alternatives
 if platform.system() == "Linux":
     config.available_features.add('linux')
+    config.available_features.add('libc++')
```
Comment (alexbatashev) [inline sycl/test/lit.cfg.py:44]: This line unconditionally disables ABI tests. Instead, it is better to add new flag [here](https://github.com/intel/llvm/blob/sycl/sycl/test/lit.site.cfg.py.in) for `@SYCL_USE_LIBCXX@` and conditionally enable `libc++` feature if the flag presents.

**Review by alexbatashev — CHANGES_REQUESTED** [review]

**Comment by bader** [comment]
@jinge90, please, add a space between tags and PR title for this and other PRs.

    [SYCL]Support ... -> [SYCL] Support

File: sycl/test/lit.cfg.py (hunk 6 above)
Comment (jinge90) [reply sycl/test/lit.cfg.py]: Hi, @alexbatashev
Just updated patch, could you help me review if it aligns with your expectation?
By the way, I changed the feature name from "libc++" to "libcxx" which aims to align with "SYCL_USE_LIBCXX".
Thanks very much.

**Review by alexbatashev — DISMISSED** [review]

File: sycl/test/lit.cfg.py (hunk 6 above)
Comment (alexbatashev) [reply sycl/test/lit.cfg.py]: Yes, it's fine now.

**Comment by jinge90** [comment]
Hi, @bader 
Could you help review this patch?
Thanks very much.

**Comment by vladimirlaz** [comment]
Shouldn't we add this build to CI? We can ad it to GitHub post-commit actions if not there yet. The following file can be updated to enable it: https://github.com/intel/llvm/blob/sycl/.github/workflows/linux_post_commit.yml

**Comment by bader** [comment]
> Hi, @bader
> Could you help review this patch?
> Thanks very much.

@jinge90, please, request review via GitHub UI:
image

I use following filter for the patches, where my review is required: "is:open is:pr review-requested:@me" and I don't see this PR because I already reviewed this patch and another review is not requested.

File: sycl/test/lit.cfg.py [hunk 7]
```diff
@@ -41,6 +41,8 @@
 # Configure LD_LIBRARY_PATH or corresponding os-specific alternatives
 if platform.system() == "Linux":
     config.available_features.add('linux')
+    if config.sycl_use_libcxx == "ON":
+        config.available_features.add('libcxx')
```
Comment (bader) [inline sycl/test/lit.cfg.py:45]: It looks like libcxx has an experimental support on Windows as well. I think LIT scripts shouldn't limit libcxx configuration to Linux platform only.

File: sycl/source/CMakeLists.txt [hunk 8]
```diff
@@ -84,7 +84,12 @@ function(add_sycl_rt_library LIB_NAME)
     PUBLIC
       $<$<BOOL:${SYCL_BUILD_PI_CUDA}>:USE_PI_CUDA>)
 
-  add_common_options(${LIB_NAME})
+  if (SYCL_USE_LIBCXX)
+    add_libcxx_compile_options(${LIB_OBJ_NAME})
+    add_libcxx_link_options(${LIB_NAME})
+  else()
+    add_common_options(${LIB_NAME})
+  endif()
```
Comment (bader) [inline sycl/source/CMakeLists.txt:92]: Is this change really necessary? Isn't `add_common_options` handles `SYCL_USE_LIBCXX` correctly?

File: sycl/CMakeLists.txt [hunk 9]
```diff
@@ -220,20 +220,28 @@ if (MSVC)
   list(APPEND SYCL_RT_LIBS sycld)
 endif()
 
+function(add_libcxx_compile_options LIB_NAME)
+  target_compile_options(${LIB_NAME} PRIVATE -nostdinc++)
```
Comment (bader) [inline sycl/CMakeLists.txt:224]: Shouldn't we use libc++ for all the targets including tools and not just runtime libraries?
If so, we probably should configure `CMAKE_CXX_FLAGS` or some other variables to set these options for all the targets.

**Comment by jinge90** [comment]
> > Hi, @bader
> > Could you help review this patch?
> > Thanks very much.
> 
> @jinge90, please, request review via GitHub UI:
> image
> 
> I use following filter for the patches, where my review is required: "is:open is:pr review-requested:@me" and I don't see this PR because I already reviewed this patch and another review is not requested.

Hi, @bader 
Thanks very much for reminder.

File: sycl/test/lit.cfg.py (hunk 7 above)
Comment (jinge90) [reply sycl/test/lit.cfg.py:45]: Hi, @bader 
I haven't tried to build sycl runtime using libcxx on Windows, @alexbatashev , Is the experimental feature(building libcxx with llvm libcxx) supported on Windows?
Thanks very much.

Comment (alexbatashev) [reply sycl/test/lit.cfg.py:45]: I have never tried it on Windows. libcxx.llvm.org does not mention Windows as one of the supported platforms, though.

Comment (bader) [reply sycl/test/lit.cfg.py:45]: I was using this document: https://libcxx.llvm.org/docs/BuildingLibcxx.html#experimental-support-for-windows

Comment (jinge90) [reply sycl/test/lit.cfg.py:45]: Hi, @bader and @alexbatashev 
If we want to enable building SYCL runtime with llvm libcxx on Windows, we need to 'use' libc++ with MSVC CL compiler on Windows which means we need to build all the SYCL runtime using MSVC CL compiler and use llvm libc++ instead of MSVC's C++ std library meanwhile.
According to the document: https://libcxx.llvm.org/docs/UsingLibcxx.html
We can find how to 'use' libc++ together with g++ or clang++ on Linux but can't find any information about using MSVC CL compiler with libc++.
Thanks very much.

File: sycl/CMakeLists.txt (hunk 9 above)
Comment (jinge90) [reply sycl/CMakeLists.txt:224]: Hi, @bader 
Previously, we wanted to investigate llvm libc++ for SYCL kernel programming and found a lot of linker error encountered when we switched to llvm libc++. After compiler team's investigation, we found building SYCL runtime with llvm libc++ can solve those linker errors. This may be why only SYCL runtime is built with llvm libc++. I remembered enabling building SYCL runtime with libc++ did cost some effort and if we want to enable building more components using llvm libc++, much more effort may be needed. @alexbatashev , please correct me if anything wrong.
If we want to use libc++ for all targets including tools per strong request, I suggest to do it in a separate PR.
Thanks very much.

File: sycl/source/CMakeLists.txt (hunk 8 above)
Comment (jinge90) [reply sycl/source/CMakeLists.txt:92]: In add_common_options, we add compile and link options to the same input parameter ${LIB_NAME}.
If using llvm libcxx to build SYCL runtime library, to all objects file, we only needed to add "-nostdinc++" in compile time and add all linker flag such as -lc++, -lc++abi in link time.
The input parameter of add_libcxx_compile_options and add_libcxx_link_options are different.

File: sycl/test/lit.cfg.py (hunk 7 above)
Comment (bader) [reply sycl/test/lit.cfg.py:45]: The point of my question is: should we block testing with libc++ on Windows in lit.cfg.py?
If so, could add a comment to the script explaining why this feature is not available on Windows platform.

File: sycl/CMakeLists.txt (hunk 9 above)
Comment (bader) [reply sycl/CMakeLists.txt:224]: > If we want to use libc++ for all targets including tools per strong request, I suggest to do it in a separate PR.

You mean abandon this patch and replace with new implementation, right?

File: sycl/source/CMakeLists.txt (hunk 8 above)
Comment (bader) [reply sycl/source/CMakeLists.txt:92]: I think we should update `add_common_options` function to accept both object and shared libraries. Current implementation doesn't seem to work for our use cases.

File: sycl/test/lit.cfg.py (hunk 7 above)
Comment (jinge90) [reply sycl/test/lit.cfg.py:48]: Hi, @bader 
I added comments to explain whey we block testing with libc++ on Windows as MSVC CL compiler doesn't support to use llvm libc++ instead of MSVC C++ std library.
Thanks very much.

File: sycl/source/CMakeLists.txt (hunk 8 above)
Comment (jinge90) [reply sycl/source/CMakeLists.txt]: Hi, @bader 
I have updated add_common_options adding one input param "IS_LINK", if IS_LINK==1, we will add some common linker option, instead, only common compiling options will be added.

File: sycl/CMakeLists.txt (hunk 9 above)
Comment (jinge90) [reply sycl/CMakeLists.txt]: Hi, @bader 
"Building the whole sycl toolchain with llvm libcxx" seems to be a 'bigger' task and needs more effort. This PR only aims to fix compiling failure issue for building SYCL runtime with llvm libcxx. If we have to enable building whole sycl toolchain with llvm libcxx per strong request, suggest to open a new PR for that.
Thanks very much.

File: sycl/CMakeLists.txt [hunk 10]
```diff
@@ -220,28 +220,25 @@ if (MSVC)
...
-  target_link_libraries(${LIB_NAME} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
-endfunction(add_libcxx_link_options)
-
 # This function allows building multiple libraries with the same options.
 # Currently used by sycl and plugins library.
 # Currently handles linking with libcxx support and gcc workaround
-function( add_common_options LIB_NAME)
+# If IS_LINK is 1, common linker options should be added for target or common
+# compile option should be added.
+function( add_common_options TAR_NAME IS_LINK)
```
Comment (bader) [inline sycl/CMakeLists.txt:228]: I think using optional parameter for setting Object library name is much cleaner interface.

Just check if additional argument is set via `ARGC` variable and if so, set compile options for Object library and link options for the first argument. Keep current behavior otherwise. This way you don't need to change multiple CMakeLitsts.txt files - just sycl/source/CMakeLists.txt.

``` CMake
add_common_options(${LIB_NAME} ${LIB_OBJ_NAME})
```

See https://cmake.org/cmake/help/latest/command/function.html#arguments for more details.

**Review by bader — CHANGES_REQUESTED** [review]

File: sycl/CMakeLists.txt [hunk 11]
```diff
@@ -223,17 +223,21 @@ endif()
 # This function allows building multiple libraries with the same options.
 # Currently used by sycl and plugins library.
 # Currently handles linking with libcxx support and gcc workaround
-function( add_common_options LIB_NAME)
+function( add_common_options)
```
Comment (bader) [inline sycl/CMakeLists.txt:226]: I don't think `add_common_options` functions should allow empty argument list.
The code below doesn't check that `ARGV0` is provided, so it might fail on wrong usages of the function.
I suggest keep `LIB_NAME` named argument.
```suggestion
function( add_common_options LIB_NAME)
```

File: sycl/CMakeLists.txt [hunk 12]
```diff
@@ -223,17 +223,21 @@ endif()
...
       if ((CMAKE_CXX_COMPILER_ID STREQUAL "GNU") OR
           (CMAKE_CXX_COMPILER_ID STREQUAL "Clang"))
-          target_compile_options(${LIB_NAME} PRIVATE -nostdinc++)
           if ((NOT (DEFINED SYCL_LIBCXX_INCLUDE_PATH)) OR (NOT (DEFINED SYCL_LIBCXX_LIBRARY_PATH)))
               message(FATAL_ERROR "When building with libc++ SYCL_LIBCXX_INCLUDE_PATHS and"
                                   "SYCL_LIBCXX_LIBRARY_PATH should be set")
           endif()
-          target_include_directories(${LIB_NAME} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
-          target_link_libraries(${LIB_NAME} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
+              target_link_libraries(${ARGV0} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
```
Comment (bader) [inline sycl/CMakeLists.txt:234]: Please, fix code indentation.

File: sycl/CMakeLists.txt [hunk 13]
```diff
@@ -223,21 +223,21 @@ endif()
...
-              target_link_libraries(${ARGV0} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
-              target_compile_options(${ARGV0} PRIVATE -nostdinc++)
-              target_include_directories(${ARGV0} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
-              if (ARGC EQUAL 2)
-                target_compile_options(${ARGV1} PRIVATE -nostdinc++)
-                target_include_directories(${ARGV1} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
-              endif()
+          target_link_libraries(${ARGV0} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
+          target_compile_options(${ARGV0} PRIVATE -nostdinc++)
+          target_include_directories(${ARGV0} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
```
Comment (bader) [inline sycl/CMakeLists.txt:236]:
```suggestion
          target_link_libraries(${LIB_NAME} PRIVATE "-L${SYCL_LIBCXX_LIBRARY_PATH}" -Wl,-rpath,${SYCL_LIBCXX_LIBRARY_PATH} -nodefaultlibs -lc++ -lc++abi -lm -lc -lgcc_s -lgcc)
          target_compile_options(${LIB_NAME} PRIVATE -nostdinc++)
          target_include_directories(${LIB_NAME} PRIVATE "${SYCL_LIBCXX_INCLUDE_PATH}")
```

**Review by bader — DISMISSED** [review]

**Review by alexbatashev — DISMISSED** [review]
